### PR TITLE
Update renovate tests for new ELF findSpaceForPHDRs return value.

### PR DESCRIPTION
**Pull request by kquick**
Updates tests for new findSpaceForPHDRs return value.

oAt present, the tests do *not* pass: the tests are modified to indicate what I believe the intended results are, but the renovate code at present does not satisfy one of the tests.  I believe this indicates an issue in the core findSpaceForPHDRs change: it's not clear that the NoAddress return value would ever be generated.

### CODE REVIEW

**Comment by travitch** [comment]
I didn't even realize that there were separate renovate tests

**Comment by langston-barrett** [comment]
> I didn't even realize that there were separate renovate tests

Pretty sure they were added with the PR that added `findSpaceForPHDRs` :smile:

**Comment by langston-barrett** [comment]
@kquick asked me to look into the test failures, here are my thoughts. You can see here

https://github.com/GaloisInc/renovate/commit/3ff83282ef4766256ca4c68b787e5361aef66107#diff-1c8442bca199478b077d30600518d77459e564576d1ab2ff72f8975287b7809dR133

and here

https://github.com/GaloisInc/renovate/commit/3ff83282ef4766256ca4c68b787e5361aef66107#diff-1c8442bca199478b077d30600518d77459e564576d1ab2ff72f8975287b7809dR138

that 3ff83282ef4766256ca4c68b787e5361aef66107 took two branches that previously returned `Nothing` and differentiated them by changing it in to `NoAddress` and `FallbackAddress`. I believe this accounts for this change:

https://github.com/GaloisInc/renovate/pull/61/files#diff-206d8cfa1c2a8f073a9aef6a1e76fc109af7084497a02c3b8bc0da7f79210f5eR53

In short, the previous code pessimistically assumed that every binary had a TLS segment, which isn't true. The `FallbackAddress` is safe to use in the case when one doesn't exist.

If this test is failing, it is likely accounted for by the same change: 

https://github.com/GaloisInc/renovate/pull/61/files#diff-206d8cfa1c2a8f073a9aef6a1e76fc109af7084497a02c3b8bc0da7f79210f5eR105

It would probably be good to construct an additional testcase that does cover the `NoAddress` case, even if it's synthetic/not from a real binary.

File: renovate/tests/Main.hs [hunk 1]
```diff
@@ -102,7 +102,7 @@ findSpaceForPHDRsTests =
         -- Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz   Flg Align
         -- LOAD           0x000000 0x00010000 0x00010000 0x0031c 0x0031c  R E 0x10000
         -- LOAD           0x00031c 0x0002031c 0x0002031c 0x00004 0x400008 RW  0x10000
-        Nothing T.@=?
+        ELF.NoAddress T.@=?
```
Comment (langston-barrett) [inline renovate/tests/Main.hs:105]: As we discussed on Mattermost, it sounds like this test is returning `FallbackAddress`, which I believe is fine. Here's how I would rewrite the above comment:
```
    -- Values in the following test case are taken from
    -- linked-list.noopt.nostdlib.aarch32.exe. The .bss section has a HUGE buffer in it, which
    -- overlaps with the optimal placement of a new PHDR segment. However, we can still find
    -- a FallbackAddress which is safe assuming the binary has no TLS segment (and it doesn't).
```

File: renovate-x86/tests/X64.hs [hunk 2]
```diff
@@ -79,10 +81,12 @@ mkTest :: C.HandleAllocator -> FilePath -> T.TestTree
 mkTest hdlAlloc fp = T.testCase fp $ withELF elfFilename testRewrite
   where
     elfFilename = replaceExtension fp "exe"
+    testRewrite :: E.SomeElf E.ElfHeaderInfo -> IO ()
     testRewrite elf = do
       Just expected <- readMaybe <$> readFile (fp <.> "expected")
-      let cfg = [(R.X86_64, R.SomeConfig NR.knownNat MBL.Elf64Repr (R64.config (R.AnalyzeOnly (analysis expected))))]
-      R.withElfConfig (E.Elf64 elf) cfg (testBlockRecovery hdlAlloc)
+      let cfg = [(R.X86_64
+                 , R.SomeConfig NR.knownNat MBL.Elf64Repr (R64.config (R.AnalyzeOnly (analysis expected))))]
```
Comment (langston-barrett) [inline renovate-x86/tests/X64.hs:88]: This looks like a formatting-only change. I like to keep formatting changes and functional changes in separate commits where possible. This makes it easier to review changes (the reviewer doesn't have to comb carefully through formatting-only changes to make sure they didn't miss something), and makes commits more focused on their intended changes.

This is just a nit/note for the future, and shouldn't hold up merging.

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs [hunk 3]
```diff
@@ -51,9 +51,12 @@ makeLoadSegmentInfo phdr =
                               , pMemSz = E.phdrMemSize phdr
                               }
 
-data PHDRAddress w = TLSSafeAddress (E.ElfWordType w)
+data PHDRAddress w = TLSSafeAddress (E.ElfWordType w)  -- TLS = Thread Local Storage
```
Comment (langston-barrett) [inline renovate/src/Renovate/BinaryFormat/ELF/Internal.hs:54]: Thanks for the note :-)

**Review by langston-barrett — APPROVED** [review]
Thanks for putting this together!

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 4]
```diff
@@ -628,7 +628,6 @@ choosePHDRSegmentAddress elf = do
   -- one can't be found.
   (tlssegment, segmentInfos) <- indexLoadableSegments elf fakePhdrSeg
   case findSpaceForPHDRs segmentInfos projectedOffset requiredSize of
-    NoAddress -> P.throwM $ NoSpaceForPHDRs (fromIntegral projectedOffset) (fromIntegral requiredSize)
```
Comment (travitch) [inline renovate/src/Renovate/BinaryFormat/ELF.hs:631]: With this removed, can the `NoSpaceForPHDRs` error still be thrown? If not, we should probably remove it.

**Review by travitch — APPROVED** [review]
Thanks - just one comment from me that could be worth addressing (or that might be irrelevant)

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 4 above)
Comment (kquick) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:631]: It can still be thrown (see line 643, about 10 down from here).

File: renovate-x86/tests/X64.hs (hunk 2 above)
Comment (kquick) [reply renovate-x86/tests/X64.hs:88]: Yes, I normally avoid that as well; I think this one just crept in accidentally.
